### PR TITLE
TimeOrderedKeyCompactionStrategy

### PR DESCRIPTION
**Pull request by gauravAshok**
Added a new table property time_ordered_key = true|false.

Example create table query:
```
CREATE TABLE grouped_msgs_sidelines
(
    ts_bucket TIMESTAMP,
    bucket_width INT,
    topic_name TEXT,
    subscription_name TEXT,
    g_id TEXT,
    PRIMARY KEY((ts_bucket, bucket_width, topic_name, subscription_name), g_id)
) WITH compaction = {
      'class' : 'org.apache.cassandra.db.compaction.TimeOrderedKeyCompactionStrategy',
      'compaction_window_unit': 'MINUTES',
      'compaction_window_size': '1',
      'split_factor':'8',
      'window_garbage_size_threshold': '128,50.0',
      'global_garbage_size_threshold': '1024,50.0'
  }
  AND time_ordered_key = true;
```

* 'time_ordered_key' flag forces the metadata collector to treat the first 2 columns of partition key of type (TIMESTAMP, INT). This allows the collector to save time window stats for each sstable.
The effect is that instead of looking into every sstable, cassandra will use the time window stats of sstable to filter out the potential sstable check resulting in much better read latency.

* We are also now collecting detailed summary of tombstones in the metadata of each sstables. Tombstones can be of 3 types: row, partition, range. This summary is used to estimate the size of data that can be freed up if compacted.

* 'split_factor' is used to limit the explosion of new & smaller files when a sstable with a really wide time window gets split.

### CODE REVIEW

File: src/java/org/apache/cassandra/io/sstable/metadata/MetadataCollector.java [hunk 1]
```diff
@@ -103,7 +100,10 @@ public static StatsMetadata defaultStatsMetadata()
     protected boolean hasLegacyCounterShards = false;
     protected long totalColumnsSet;
     protected long totalRows;
+    protected MinMaxLongTracker keyRangeTracker = new MinMaxLongTracker(Long.MAX_VALUE, Long.MIN_VALUE);
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/io/sstable/metadata/MetadataCollector.java:102]: The constructor is (defaultMin, defaultMax). Any reason you're using long.max for defaultMin instead of the other way? Won't affect anything as long as the tracker is updated at least once before being queried for min/max

File: src/java/org/apache/cassandra/db/compaction/writers/TimeBasedSplittingCompactionWriter.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.cassandra.db.rows.UnfilteredRowIterator;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.sstable.format.SSTableWriter;
+import org.apache.cassandra.io.sstable.metadata.MetadataCollector;
+
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+public class TimeBasedSplittingCompactionWriter extends CompactionAwareWriter
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/writers/TimeBasedSplittingCompactionWriter.java:36]: To confirm, what this is doing is:
* Split given sstables into new sstables
* How many sstables to create is defined by splitFactor argument
* For given row, we look at which window it falls in (using it's timewindow, the windowSize, the windowStart), and then the sstable. Essentially the duration [windowStart, windowStart + windowSize*windowCount] is partitioned into splitfactor sub-ranger equally.

Assuming my understanding is right, why is windowStart, windowCount manually specified? Isn't it better to look at the data in the sstables and figure out how many windows it covers, and the starting window (the oldest one among all)?

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 3]
```diff
@@ -0,0 +1,946 @@
...
+                return null;
+            }
+
+            LifecycleTransaction modifier = cfs.getTracker().tryModify(candidate.sstables, OperationType.COMPACTION);
+            if (modifier != null)
+            {
+                logger.debug("{}: {}", tableName, candidate);
+                return buildCompactionTask(candidate, modifier, gcBefore);
+            }
+            previousCandidate = candidate;
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:127]: From what I can see, we get here only if we fail to get txn from tryModify.
In that case why do we rerun the logic to get sstables to compact? Why not retry with same sstables?

File: src/java/org/apache/cassandra/db/compaction/writers/TimeBasedSplittingCompactionWriter.java (hunk 2 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/writers/TimeBasedSplittingCompactionWriter.java:36]: yup, something i didnt realise.
Figuring out the range of the windows, is something that is done by another class which creates this class's object.
But yes, it does make sense to put that logic here only.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 3 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:127]: Same sstables would not make sense here. 
tryModify will see if the candidate is eligible for compaction atomically. If not, it returns null. In such a case we store the candidate, and rerun the compaction to get a fresh candidate. If the new candidate is again equal to the previous candidate, then give up. CompactionManager will take care of calling the strategy again when any previously running compaction finishes or a memtable flush happens.
compaction strategy can run in parallel to the compaction itself (to get the more parallel compaction tasks.), but if the compaction strategy just gets the same thing again and again, just stop it.

File: src/java/org/apache/cassandra/io/sstable/metadata/MetadataCollector.java (hunk 1 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/io/sstable/metadata/MetadataCollector.java:102]: As a default state, which could easily be checked for if needed.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 4]
```diff
@@ -0,0 +1,950 @@
...
+                    ", stats=" + stats +
+                    '}';
+        }
+    }
+
+    CategorizedSSTables categorizeSStables(Iterable<SSTableReader> sstables, int gcBefore)
+    {
+        int gcGraceSecondsMs = Math.max(1, cfs.metadata.params.gcGraceSeconds) * 1000;
+        long nowMs = System.currentTimeMillis();
+        long currentGcWindowMs = toWindowMs(nowMs, gcGraceSecondsMs);
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Is the window size for compaction always same as gc grace seconds?
Also gc grace seconds can be zero right? (Even though we won't have that for our usage)

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 5]
```diff
@@ -0,0 +1,950 @@
...
+
+        OverallStats overallStats = new OverallStats();
+        sstables.forEach(s -> {
+            SSTableStats stats = getOrCompute(s);
+            collect(overallStats, stats);
+            if (s.getSSTableLevel() == SSTable.TOMBSTONE_SSTABLE_LVL)
+            {
+                // If gcGraceSeconds > 0, i.e. tombstones will linger in sstables for some time and if they are created recently,
+                // mark them as latest so that they are not compacted against data sstables just yet.
+                if (cfs.metadata.params.gcGraceSeconds > 0 && toWindowMs(s.getSSTableMetadata().maxLocalDeletionTime * 1000, gcGraceSecondsMs) >= currentGcWindowMs)
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: I'm not clear why we're considering both gcBefore and the current time (currentGcWindowMs). Shouldn't it be entirely based on gcBefore? My understanding is that the reason for gcBefore param is that all methods have unified view of what the cutoff point is for gc, instead of everyone calling now() and getting inconsistent results.

IMO from what I understand of the categorization, we should just figure out the window corresponding to gcBefore, then if window of maxLocalDeletion time of table is:
```
> gcWindow => latest
= gcWindow => nearExpiry (will become expirable in next window)
< gcWindow => compactable 
```

Currently I'm finding it's hard to reason exactly when a table is in which category from the code.

File: src/java/org/apache/cassandra/utils/TimeWindow.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Time window representing a range [ts, ts + duration). Use it with ms only to avoid any bugs.
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/utils/TimeWindow.java:25]: Maybe add ms to field names to make that explicit.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 7]
```diff
@@ -0,0 +1,950 @@
...
+            if (this.cardinality != null)
+            {
+                this.keyCount = cardinality.cardinality();
+            }
+            else
+            {
+                this.keyCount = sstable.estimatedKeys();
+            }
+            this.rowCount = sstable.getTotalRows();
+            this.timeWindowMs = new TimeWindow(meta.minKey, (int) (meta.maxKey - meta.minKey));
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:470]: Assuming metadata is written so that maxKey is timestamp of oldest key + 1, since TimeWindow is like [start,end)

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 8]
```diff
@@ -0,0 +1,950 @@
...
+    {
+        if (ss.sstable.getSSTableLevel() == SSTable.DATA_SSTABLE_LVL)
+        {
+            os.keyCount += ss.keyCount;
+            os.rowCount += ss.rowCount;
+            os.onDiskLength += ss.sstable.onDiskLength();
+        }
+        else
+        {
+            os.tombstoneCounts.add(ss.tombstoneCounts);
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:531]: Why do we exclude tombstone tables from onDiskLength?

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 9]
```diff
@@ -0,0 +1,950 @@
...
+    {
+        double avgPartitionSize = totalDataPartitions == 0 ? 0 : (double) totalDataSizeOnDisk / totalDataPartitions;
+        double avgRowSize = totalDataRows == 0 ? 0 : (double) totalDataSizeOnDisk / totalDataRows;
+
+        long partitionTombstones = Math.min(tombstoneCounts.partitionTombstones, totalDataPartitions);
+        long rowTombstones = Math.min(tombstoneCounts.rowTombstones, totalDataRows);
+
+        // we are over eastimating the garbage. so assuming that range tombstones are for the remaining undeleted partitions.
+        long rangeTombstones = Math.min(tombstoneCounts.rangeTombstones / 2, totalDataPartitions - partitionTombstones);
+        return (partitionTombstones * avgPartitionSize) + (rowTombstones * avgRowSize) + (rangeTombstones * avgPartitionSize / 3.0);
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:204]: Any particular reasoning for factor of 2 and 3 for range tombstones or just a guesstimate?

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 10]
```diff
@@ -0,0 +1,950 @@
...
+        if (maxGarbageTombstones.sstables.size() + maxGarbageData.size() > getMaxFileCountForCompaction() || dataSSTableTimeWindowMs.duration > 2 * getWindowSizeInMs())
+        {
+            if (maxGarbageTombstones.timeWindowMs.duration > 2 * getWindowSizeInMs() || maxGarbageTombstones.sstables.size() > 1)
+            {
+                logger.debug("{}: too many tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(maxGarbageTombstones.sstables), true, true));
+            }
+
+            List<OverlappingSet> overlappingDataSets = getDistinctOverlappingSSTables(maxGarbageData);
+            long windowSzInMs = getWindowSizeInMs();
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Do we want to support changing compaction window size dynamically?

Practically speaking, changing it for a live db will lead to huge amounts of compaction as the strategy tries to break-up and merge all the sstables to re-align with new time window boundaries.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 11]
```diff
@@ -0,0 +1,950 @@
...
+
+            logger.debug("{}: too many data files", tableName);
+            // we are here, which means that tombstone.timerange <= 2 * window && tombstone.size() <= 1, and there could be too many data files.
+            // data files could be too wide or many small files.
+            // we would like to merge all those files that are overlapping or are in a single compaction window.
+            return overlappingDataSets.stream()
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Consider refactoring this, feels a bit hard to follow what's happening, especially with the multiple single character temp vars (s, v, l) and the types (we're dealing with Map of List of Overlapping Set of SSTable from what I can see). Might be better to just replace with basic for loops.

Just validate my understanding:
For all the data tables, we group them by compaction window. All data tables of a particular overlapping set belong to the window corresponding to the min. start timestamp of all the tables of the overlapping set.
Then we just pick the window with the largest number of data tables. 
We don't want to compact too many of them at once, so we limit to maxFileCountForCompaction.

Also in general this method seems to be too complex and long, maybe break it apart into small ones. Like a separate method to handle the case of too many files

Comment (hoodakaushal) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Also I'd invert the if condition. 
```
Instead of 
find candidate
if(candidate too large){
do different stuff
} else{
return candidate
}

maybe do:

if(candidate not too large){
return candidate
}
do different stuff
```

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 12]
```diff
@@ -0,0 +1,950 @@
...
+                    start = Long.MAX_VALUE;
+                    end = Long.MIN_VALUE;
+                    currentSet.maxOverlap = -1;
+                }
+            }
+        }
+        return sets;
+    }
+
+    static List<SSTableStats> getMaxOverlappingSSTables(List<SSTableStats> sstables)
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: This is the same logic as the distinct one above right? Maybe just use that and return the largest of hte overlapping sets..

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 13]
```diff
@@ -0,0 +1,950 @@
...
+            return Optional.of(new SSTablesForCompaction(all, false, false));
+        }
+    }
+
+    private Optional<SSTablesForCompaction> getCompactionCandidateBasedOnOverlaps(CategorizedSSTables categorizedSSTables)
+    {
+        int maxFileForCompaction = getMaxFileCountForCompaction();
+        List<SSTableStats> mergeableLatestTombstones = getMaxOverlappingSSTables(categorizedSSTables.latestTombstones);
+        // if latest tombstones are overlapping too much, merge them without splitting.
+        if (mergeableLatestTombstones.size() > 1)
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Why the higher priority for different types of sstable? If the idea is simply to reduce overlap, shouldn't we pick the largest overlappingSet across all categories? (data, expired, latest etc)

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 14]
```diff
@@ -0,0 +1,950 @@
...
+        }
+
+        // If there are too many files OR (data size is over the threshold and the data set is too wide),
+        // we need a sure shot way of reducing them.
+        if (maxGarbageTombstones.sstables.size() + maxGarbageData.size() > getMaxFileCountForCompaction() || dataSSTableTimeWindowMs.duration > 2 * getWindowSizeInMs())
+        {
+            if (maxGarbageTombstones.timeWindowMs.duration > 2 * getWindowSizeInMs() || maxGarbageTombstones.sstables.size() > 1)
+            {
+                logger.debug("{}: too many tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(maxGarbageTombstones.sstables), true, true));
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Considering tombstoneMerge = true => the expired tombstones will not be present in resulting sstable(s),
and that these are ALL garbage tombstones, looks to me like we'll just delete a bunch of tombstones without deleting the corresponding data.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 15]
```diff
@@ -0,0 +1,950 @@
...
+                    // limit the count of sstables to maxFileCountForCompaction
+                    .map(l -> limit(getMaxFileCountForCompaction(), l))
+                    // create a compaction task
+                    .map(l -> new SSTablesForCompaction(toSSTableReader(l), false, true));
+        }
+        else
+        {
+            List<SSTableReader> all = new ArrayList<>(toSSTableReader(maxGarbageTombstones.sstables));
+            all.addAll(toSSTableReader(maxGarbageData));
+            return Optional.of(new SSTablesForCompaction(all, false, false));
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: And here tombstoneMerge = false means that we will delete the data but keep the corresponding tombstones, and the writer will presumably put them back out as new tombstone tables.
So we have two types of expired tombstones - either whose corresponding data is deleted, in which case they can (and should) be GCd, or whose corresponding data is not deleted, and they should not be GCd. 

I don't see how this is handled. From what I understand, we'll keep piling up these double garbage (expired and corresponding data is gone) tombstone tables, and eventually merge them due to too many files or something, but that might also result in GC of some tombstones that we want to keep, because their data is still not deleted.

EDIT: LOL ignore this, I misunderstood tombstoneMerge. If it's TRUE we DON'T delete expired tombstones.

Comment (hoodakaushal) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Why is splitSSTable false here? Considering that the overlap range can be huge, don't we want to split the resulting data?

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategyOptions.java [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+    final TimeWindowCompactionStrategyOptions twcsOptions;
+
+    final long windowCompactionSizeInMB;
+    final double windowCompactionSizePercent;
+
+    final long windowCompactionGlobalSizeInMB;
+    final double windowCompactionGlobalSizePercent;
+
+    final int splitFactor;
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategyOptions.java:49]: What's the significant of splitFactor?

From what I can see, it is used eventually by TBSCWriter. But I thought the idea was to split SSTables aligning with time window boundaries, and this will just split them into equal time ranges. 

Edit: read the description in the PR, it's to stop explosion of smaller files. But given the strategy will eventually anyway break everything down into window level sstables, this doesn't seem to prevent that. Seems to me instead of creating 100 tables at start, we can use split factor to limit them to 10, but the compaction strategy will again split them up eventually to 100.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 17]
```diff
@@ -0,0 +1,950 @@
...
+     * 1. Time overlap between sstables. Many files overlaping with each other -> High read latency.
+     * 2. Time range of a sstable. Large range -> large size. Likelihood of it overlapping with other sstables also increases when the range is large, which
+     * will lead to larger cost in compaction. Compaction will be most efficient if there are fewer files to compact at a time.
+     * <p>
+     * This compaction strategy aims to reduce the max overlap and the number of sstables.
+     *
+     * @param gcBefore
+     * @return
+     */
+    synchronized SSTablesForCompaction getNextBackgroundSSTables(final int gcBefore)
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:151]: Why are all these synchronized? Are multiple threads going to be running/getting compaction tasks? Is a compaction strategy expected to be threadsafe?

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java [hunk 18]
```diff
@@ -0,0 +1,950 @@
...
+            {
+                dWindowOccupancy += collectOccupacyStat(dWindowSet, windowSizeMs, tw);
+            }
+            else
+            {
+                tWindowOccupancy += collectOccupacyStat(tWindowSet, windowSizeMs, tw);
+            }
+        }
+
+        int n = tWindowOccupancy - (2 * tWindowSet.size()) + dWindowOccupancy - (2 * dWindowSet.size());
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Can you explain this?

File: src/java/org/apache/cassandra/db/lifecycle/LogFile.java [hunk 19]
```diff
@@ -60,7 +60,7 @@
     static String EXT = ".log";
     static char SEP = '_';
     // cc_txn_opname_id.log (where cc is one of the sstable versions defined in BigVersion)
-    static Pattern FILE_REGEX = Pattern.compile(String.format("^(.{2})_txn_(.*)_(.*)%s$", EXT));
+    static Pattern FILE_REGEX = Pattern.compile(String.format("^([a-z]+)_txn_(.*)_(.*)%s$", EXT));
```
Comment (hoodakaushal) [inline src/java/org/apache/cassandra/db/lifecycle/LogFile.java:63]: So this is because you changed sstable metadata version to mdn right?

File: src/java/org/apache/cassandra/utils/TimeWindow.java (hunk 6 above)
Comment (hoodakaushal) [reply src/java/org/apache/cassandra/utils/TimeWindow.java:25]: Also would it make sense to use Interval instead. It's not in the start, duration style, but still..

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 7 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:470]: in the metadataCollector you can see that the maxKey is calculated from timestamp in partition key + the bucket size (again present in partition key).
Its not +1. if the granularity of time bucket in PK is 250 ms, maxKey will be newest key + 250 ms.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 8 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:531]: On first thought it is just easier to assume that tombstones have negligible data size.
In my testing, 240K tombstones resulted in few MB of data.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 9 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:204]: 2 is coming from the fact that each range tombstone results in 2 tombstone markers. 1 for the start and another for the end.
3 is coming from the fact that average length of an interval when start & end are chosen randomly from (a,b) is (b-a) / 3.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 10 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Actually, i can put a unit test for this. But what i expect that unnecessarily this wont happen.
Merging and split only happens when there is an overlap or compaction running. So basically whenever the compaction runs, it will start aligning to new window. 
Also once files have been split, there should be really low overlap. Hence increase window size will start merging them and decreasing window size will start splitting them; only on the next compaction.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 12 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: already refactored. Its not really the largest of the overlapping sets.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 15 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Already changed split to true. Initial thought was that if we are able to compact them together in whatever condition they are, let them be. 
But in one of the test it didnt looked that good, since it led to more compaction iterations.
Split = true removes 1 extra compaction, since we are compacting anyway, just use this opportunity to split them too.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategyOptions.java (hunk 16 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategyOptions.java:49]: Its a heuristic. It still needs to be seen how it will perform in actual workload. 
Though the idea is that, splitting partially first may lead to compaction of one of the split tombstones first. I am relying on the fact that tombstones wont be uniformly distributed across time range. hence splitting by some factor, i am looking for that time range which is loaded with tombstones, which if other constraint succeed, will lead to compaction. Other files dont need to be split if they dont have enough garbage.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 17 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:151]: probably for a safety. Other classes also have it. I havent seen it running in parallel.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 18 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: even i dont know this :p 
This method needs to be reworked and i am looking for some clever way to calculate the pending tasks.

File: src/java/org/apache/cassandra/db/lifecycle/LogFile.java (hunk 19 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/lifecycle/LogFile.java:63]: yes

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 4 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: yes, gc grace seconds can be zero. 
Math.max(1, cfs.metadata.params.gcGraceSeconds) is kinda useless now. As when i am using the variable "gcGraceSecondsMs", i am explicitly checking for gcgraceSeconds to be nonZero.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 5 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: We cannot use gcBefore instead of currentGcWindowMs.
If I do that, there will be 2 categories then. 
<------------------->|<--------->
          expired                     latest
And I intend to merge all the latest tombstones together. In which case the merged file can never get expired. Note that the tombstone sstable expires when its latest delete timestamp < gcBefore. But merging latest sstables may not allow that to happen.

So, i went with a fix window and created 3 categories.
<------------------->|<--->|<-------->
        expired              near expiry   latest
The latest window is fixed, and we will merge whatever sstables lie in that window. As soon as that window passes, they will fall into "near expiry" window and soon into expired window.

And about the first if check, the check gcGraceSeconds > 0 means that if gcGraceSeconds = 0 then the sstable will never be latest, they are expired as soon as they are created. To be honest the same check should be there for "near expiry" also.

File: src/java/org/apache/cassandra/utils/TimeWindow.java (hunk 6 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/utils/TimeWindow.java:25]: interval as in [start,end) instead of [start, start + duration)?

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 11 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: You have correct understanding.
Normal loops were introducing some really large types. streams looked a bit clear.
Yes the method in itself is quite large at the moment and can be split.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 13 above)
Comment (gauravAshok) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java]: Its because tombstones files move from latest -> near expiry -> expired category.
all having different characteristcs.
Latest will have new tombstones sstables on every flush and will probably be wide. So we keep merging them to keep number of files in this category low on the first opportunity.

Near expiry is a transition period. near sstables here are added every gcSeconds. It should have low overlap. But if they are wide (as they will be, the result of merging in latest sstables.), split them now so that when they get expired they can be efficiently merged together with the data files.
Then in the last, take care of the all files in the expired tombstones and data files.

File: src/java/org/apache/cassandra/utils/TimeWindow.java (hunk 6 above)
Comment (hoodakaushal) [reply src/java/org/apache/cassandra/utils/TimeWindow.java:25]: Yes.

File: src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java (hunk 8 above)
Comment (hoodakaushal) [reply src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java:531]: Makes sense.

**Review by hoodakaushal — COMMENTED** [review]
LGTM
